pkg/monitor/agent_test.go

```diff
@@ -0,0 +1,167 @@
+		assert.NoError(t, err)
+		for i := range partiallyUpdatedCRD.AgentConditions {
+			partiallyUpdatedCRD.AgentConditions[i].LastHeartbeatTime.Time = crd.AgentConditions[i].LastHeartbeatTime.Time
+		}
+		assert.Equal(t, partiallyUpdatedCRD, crd)
+	})
+	t.Run("partial update-failure", func(t *testing.T) {
+		clientset := fakeclientset.NewSimpleClientset(existingCRD)
+		clientset.PrependReactor("update", "antreaagentinfos", func(action cgtesting.Action) (handled bool, ret runtime.Object, err error) {
+			return true, &v1beta1.AntreaAgentInfo{}, errors.New("error updating agent crd")
+		})
+		monitor := newAgentMonitor(clientset, t)
+		monitor.agentCRD = existingCRD
+		monitor.syncAgentCRD()
+		_, err := monitor.client.CrdV1beta1().AntreaAgentInfos().Update(context.TODO(), existingCRD, metav1.UpdateOptions{})
+		assert.Equal(t, "error updating agent crd", err.Error())
+		assert.Nil(t, monitor.agentCRD)
+	})
+	t.Run("get-failure", func(t *testing.T) {
+		clientset := fakeclientset.NewSimpleClientset()
+		clientset.PrependReactor("get", "antreaagentinfos", func(action cgtesting.Action) (handled bool, ret runtime.Object, err error) {
+			return true, &v1beta1.AntreaAgentInfo{}, errors.New("error getting agent crd")
+		})
+		monitor := newAgentMonitor(clientset, t)
+		monitor.agentCRD = existingCRD
+		monitor.syncAgentCRD()
+		_, err := monitor.client.CrdV1beta1().AntreaAgentInfos().Get(context.TODO(), "testAgentCRD", metav1.GetOptions{})
+		assert.Equal(t, "error getting agent crd", err.Error())
+		assert.Nil(t, monitor.agentCRD)
+	})
+	t.Run("entire update-success", func(t *testing.T) {
+		clientset := fakeclientset.NewSimpleClientset(existingCRD)
+		monitor := newAgentMonitor(clientset, t)
+		monitor.querier.GetAgentInfo(entirelyUpdatedCRD, false)
+		monitor.syncAgentCRD()
+		crd, err := monitor.client.CrdV1beta1().AntreaAgentInfos().Get(context.TODO(), "testAgentCRD", metav1.GetOptions{})
+		assert.NoError(t, err)
+		for i := range partiallyUpdatedCRD.AgentConditions {
+			entirelyUpdatedCRD.AgentConditions[i].LastHeartbeatTime.Time = crd.AgentConditions[i].LastHeartbeatTime.Time
+		}
+		assert.Equal(t, entirelyUpdatedCRD, crd)
+	})
+	t.Run("entire update-failure", func(t *testing.T) {
+		clientset := fakeclientset.NewSimpleClientset(existingCRD)
+		clientset.PrependReactor("update", "antreaagentinfos", func(action cgtesting.Action) (handled bool, ret runtime.Object, err error) {
+			return true, &v1beta1.AntreaAgentInfo{}, errors.New("error updating agent crd")
+		})
+		monitor := newAgentMonitor(clientset, t)
+		monitor.syncAgentCRD()
+		_, err := monitor.client.CrdV1beta1().AntreaAgentInfos().Update(context.TODO(), existingCRD, metav1.UpdateOptions{})
+		assert.Equal(t, "error updating agent crd", err.Error())
+		assert.Nil(t, monitor.agentCRD)
+	})
+}
+
+func newAgentMonitor(crdClient *fakeclientset.Clientset, t *testing.T) *agentMonitor {
+	client := fake.NewSimpleClientset()
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
```

Comment:
this is a bit weird. It feels like `ctrl.Finish()` should be called at the end of each individual test, not when this function returns.

The good news is you don't actually need to call `ctrl.Finish` any more. From the gomock documentation:
> New in go1.14+, if you are passing a *testing.T into NewController function you no longer need to call ctrl.Finish() in your test methods.

So you should be able to remove this defer statement.

Reply:
fixed